pilot/pkg/xds/sds.go

```diff
@@ -0,0 +1,291 @@
+		}
+		return SecretResource{Type: KubernetesSecretType, Name: name, Namespace: namespace, ResourceName: resource}, nil
+	}
+	return SecretResource{}, fmt.Errorf("unknown resource type: %v", resource)
+}
+
+func needsUpdate(proxy *model.Proxy, updates model.XdsUpdates) bool {
+	if proxy.Type != model.Router {
+		return false
+	}
+	if len(updates) == 0 {
+		return true
+	}
+	if len(model.ConfigNamesOfKind(updates, collections.K8SCoreV1Secrets.Resource().GroupVersionKind())) > 0 {
+		return true
+	}
+	return false
+}
+
+func (s *SecretGen) Generate(proxy *model.Proxy, _ *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	if !needsUpdate(proxy, updates) {
+		return nil
+	}
+	results := model.Resources{}
+	for _, resource := range w.ResourceNames {
+		sr, err := parseResourceName(resource, proxy.ConfigNamespace)
+		if err != nil {
+			adsLog.Warnf("error parsing resource name: %v", err)
+			continue
+		}
+
+		isCAOnlySecret := strings.HasSuffix(sr.Name, GatewaySdsCaSuffix)
+		if isCAOnlySecret {
+			secret := getCaCert(s.secrets, sr)
+			if secret != nil {
+				results = append(results, util.MessageToAny(secret))
+			}
+		} else {
+			secret := getKeyAndCert(s.secrets, sr)
+			if secret != nil {
+				results = append(results, util.MessageToAny(secret))
+			}
+		}
+	}
+	return results
+}
+
+func getCaCert(secrets informersv1.SecretInformer, res SecretResource) *tls.Secret {
+	strippedName := strings.TrimSuffix(res.Name, GatewaySdsCaSuffix)
+	k8sSecret, err := secrets.Lister().Secrets(res.Namespace).Get(strippedName)
+	var rootCert []byte
+	if err != nil {
+		// Could not fetch cert, look for legacy secret with -cacert suffix
+		k8sSecret, caCertErr := secrets.Lister().Secrets(res.Namespace).Get(res.Name)
+		if caCertErr != nil {
+			log.Warnf("requested secret %v/%v ca certificate not found: %v, %v", res.Namespace, strippedName, err, caCertErr)
+			return nil
+		}
+		rootCert = extractRoot(k8sSecret)
+	} else {
+		rootCert = extractRoot(k8sSecret)
+		// Secret exists, but does not have the ca cert. Fall back to -cacert secret
+		if rootCert == nil {
+			k8sSecret, caCertErr := secrets.Lister().Secrets(res.Namespace).Get(res.Name)
+			if caCertErr != nil {
+				log.Warnf("requested secret %v/%v ca certificate not found: %v", res.Namespace, strippedName, caCertErr)
+				return nil
+			}
+			rootCert = extractRoot(k8sSecret)
+		}
+	}
+	if rootCert == nil {
+		log.Warnf("could not find ca certificate for %v/%v", res.Namespace, strippedName)
+	}
+
+	return &tls.Secret{
+		Name: res.ResourceName,
+		Type: &tls.Secret_ValidationContext{
+			ValidationContext: &tls.CertificateValidationContext{
+				TrustedCa: &core.DataSource{
+					Specifier: &core.DataSource_InlineBytes{
+						InlineBytes: rootCert,
+					},
+				},
+			},
+		},
+	}
+}
+
+func getKeyAndCert(secrets informersv1.SecretInformer, res SecretResource) *tls.Secret {
+	k8sSecret, err := secrets.Lister().Secrets(res.Namespace).Get(res.Name)
```

Comment:
Should we do the same separation as the existing code (k8s stuff in kube/, intermediate model with push.GetSecrets(), etc.. and then the discovery code consuming the push.. Will allow us to implement other secret sources

Reply:
Yeah I think so, for now I just piled it all in here but makes sense to split it out before merging

Reply:
we can split it later - good idea, but not blocking this PR.

Reply:
I already did this